server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -56,12 +72,24 @@
  */
 public class ShardsAvailabilityHealthIndicatorService implements HealthIndicatorService {
 
+    private static final Logger LOGGER = LogManager.getLogger(ShardsAvailabilityHealthIndicatorService.class);
+
     public static final String NAME = "shards_availability";
 
+    private static final String DATA_TIER_ALLOCATION_DECIDER_NAME = "data_tier";
+
     private final ClusterService clusterService;
+    private final AllocationService allocationService;
+    private final AllocationDeciders allocationDeciders;
```

Comment:
Would it make sense to have this diagnosis extracted into a service of its own?

Reply:
This is an interesting idea. I'm inclined to disagree for now because we only make use of this logic in the health indicator whereas I think a separate service should be more reusable. That said, the indicator code is getting filled up with diagnosis methods and getting harder to read. Maybe it makes sense to do this as a follow up so it's easier to throw out if we don't like how it looks?